crates/driver/src/infra/solver/protocol_fees.rs

```diff
@@ -0,0 +1,270 @@
+use {
+    crate::{
+        domain::{
+            competition,
+            competition::{
+                order,
+                order::{fees, Side},
+                solution::{trade::ClearingPrices, Trade},
+                Solution,
+            },
+            eth,
+        },
+        infra::{
+            config::file::{FeeHandler, ProtocolFeeDiscountKind},
+            solver::{
+                dto::{FeePolicy, Order},
+                solution::fee,
+            },
+        },
+    },
+    primitive_types::H160,
+    std::{collections::HashMap, sync::Arc},
+    thiserror::Error,
+};
+
+pub type ProtocolFeeDiscountedAddresses = HashMap<H160, ProtocolFeeDiscountKind>;
+
+/// `ProtocolFees` contains all the configurable logic related to the solver
+/// protocol fees
+#[derive(Clone, Debug)]
+pub struct ProtocolFees {
+    /// Determines whether the `solver` or the `driver` handles the fees
+    fee_handler: FeeHandler,
+    /// List of addresses which receives a specific discount for the protocol
+    /// fees
+    protocol_fee_discounted_addresses: Arc<ProtocolFeeDiscountedAddresses>,
```

Comment:
I made this an `Arc<>` because the parent struct of `ProtocolFees` implements `Clone`, and I don't want to make a full copy of this `HashMap<>` every time `.clone()` is called. Wrapping it into an `Arc<>` avoids making hard copies of it.